src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -499,3 +500,77 @@ auto mf::NullWlSurfaceRole::scene_surface() const -> std::optional<std::shared_p
 void mf::NullWlSurfaceRole::refresh_surface_data_now() {}
 void mf::NullWlSurfaceRole::commit(WlSurfaceState const& state) { surface->commit(state); }
 void mf::NullWlSurfaceRole::surface_destroyed() {}
+
+mf::DragWlSurfaceRole::DragWlSurfaceRole(WlSurface* surface)
+    : NullWlSurfaceRole(surface),
+      surface{surface}
+{
+    surface->set_role(this);
+    create_scene_surface();
+}
+
+mf::DragWlSurfaceRole::~DragWlSurfaceRole()
+{
+    if (surface)
+    {
+        surface.value().clear_role();
+
+        if (shared_scene_surface)
+        {
+            auto const& session = surface.value().session;
+            session->destroy_surface(shared_scene_surface);
+            shared_scene_surface.reset();
+        }
+    }
+}
+
+auto mf::DragWlSurfaceRole::scene_surface() const -> std::optional<std::shared_ptr<scene::Surface>>
+{
+    if (shared_scene_surface)
+    {
+        return shared_scene_surface;
+    }
+    else
+    {
+        return std::nullopt;
+    }
```

Comment:
You're referring to `surface` and not `shared_scene_surface`, right? Because this seems right for the former but the latter isn't created until `DragWlSurfaceRole::create_scene_surface()`.